File: ios/FilesBridge.m
#import "FilesBridge.h"

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h”
#endif

#import <Textile/TextileApi.h>
#import "utils.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@implementation FilesBridge

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue {
  return dispatch_queue_create("io.textile.TextileNodeQueue", DISPATCH_QUEUE_SERIAL);
}

RCT_EXPORT_METHOD(addData:(NSString *)base64 threadId:(NSString *)threadId caption:(NSString *)caption resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [Textile.instance.files addData:base64 threadId:threadId caption:caption completion:^(Block * _Nullable block, NSError * _Nonnull error) {
    fulfillWithResult([block.data base64EncodedStringWithOptions:0], error, resolve, reject);
  }];
}

RCT_EXPORT_METHOD(addFiles:(NSString *)paths threadId:(NSString *)threadId caption:(NSString *)caption resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [Textile.instance.files addFiles:paths threadId:threadId caption:caption completion:^(Block * _Nullable block, NSError * _Nonnull error) {
    fulfillWithResult([block.data base64EncodedStringWithOptions:0], error, resolve, reject);
  }];
}

RCT_EXPORT_METHOD(shareFiles:(NSString *)target threadId:(NSString *)threadId caption:(NSString *)caption resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [Textile.instance.files shareFiles:target threadId:threadId caption:caption completion:^(Block * _Nullable block, NSError * _Nonnull error) {
    fulfillWithResult([block.data base64EncodedStringWithOptions:0], error, resolve, reject);
  }];
}

RCT_EXPORT_METHOD(file:(NSString*)blockId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    NSError *error;
    Files *files = [Textile.instance.files file:blockId error:&error];
    fulfillWithResult([files.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(list:(NSString*)threadId offset:(NSString*)offset limit:(NSInteger)limit resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  FilesList *list = [Textile.instance.files list:threadId offset:offset limit:limit error:&error];
  fulfillWithResult([list.data base64EncodedStringWithOptions:0], error, resolve, reject);
}

RCT_EXPORT_METHOD(content:(NSString*)hash resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [Textile.instance.files content:hash completion:^(NSData * _Nullable data, NSString * _Nullable mediaType, NSError * _Nonnull error) {
      if (data && !error) {
          NSDictionary *result = @{
              @"data" : [data base64EncodedStringWithOptions:0],
              @"mediaType" : mediaType
          };
          fulfillWithResult(result, error, resolve, reject);
      } else {
          fulfillWithResult(nil, error, resolve, reject);
      }
  }];
}

RCT_EXPORT_METHOD(imageContentForMinWidth:(NSString*)pth minWidth:(NSInteger)minWidth resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [Textile.instance.files imageContentForMinWidth:pth minWidth:minWidth completion:^(NSData * _Nullable data, NSString * _Nullable mediaType, NSError * _Nonnull error) {
      if (data && !error) {
          NSDictionary *result = @{
              @"data" : [data base64EncodedStringWithOptions:0],
              @"mediaType" : mediaType
          };
          fulfillWithResult(result, error, resolve, reject);
      } else {
          fulfillWithResult(nil, error, resolve, reject);
      }
  }];
}

@end

@implementation RCTBridge (FilesBridge)

- (FilesBridge *)filesBridge {
  return [self moduleForClass:[FilesBridge class]];
}

@end

File: android/src/main/java/io/textile/rnmobile/FilesBridge.java
package io.textile.rnmobile;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.textile.pb.Model;
import io.textile.pb.View;
import io.textile.textile.Handlers;
import io.textile.textile.Textile;

public class FilesBridge extends ReactContextBaseJavaModule {

    private Executor executor = Executors.newSingleThreadExecutor();

    public FilesBridge(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "FilesBridge";
    }

    @ReactMethod
    public void addData(final String base64, final String threadId, final String caption, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                Textile.instance().files.addData(
                        base64, threadId, caption, new Handlers.BlockHandler() {
                            @Override
                            public void onComplete(final Model.Block block) {
                                promise.resolve(Util.encode(block.toByteArray()));
                            }

                            @Override
                            public void onError(final Exception e) {
                                promise.reject("addData", e);
                            }
                        });
            }
        });
    }

    @ReactMethod
    public void addFiles(final String files, final String threadId, final String caption, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                Textile.instance().files.addFiles(files, threadId, caption, new Handlers.BlockHandler() {
                    @Override
                    public void onComplete(final Model.Block block) {
                        promise.resolve(Util.encode(block.toByteArray()));
                    }

                    @Override
                    public void onError(final Exception e) {
                        promise.reject("addFiles", e);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void shareFiles(final String hash, final String threadId, final String caption, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                Textile.instance().files.shareFiles(hash, threadId, caption, new Handlers.BlockHandler() {
                    @Override
                    public void onComplete(final Model.Block block) {
                        promise.resolve(Util.encode(block.toByteArray()));
                    }

                    @Override
                    public void onError(final Exception e) {
                        promise.reject("shareFiles", e);
                    }
                });
            }
        });
    }

    @ReactMethod
    public  void file(final String blockId, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    final View.Files files = Textile.instance().files.file(blockId);
                    promise.resolve(Util.encode(files.toByteArray()));
                }
                catch (final Exception e) {
                    promise.reject("file", e);
                }
            }
        });
    }

    @ReactMethod
    public void list(final String threadId, final String offset, final Integer limit, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    final View.FilesList list = Textile.instance().files.list(threadId, offset, limit);
                    promise.resolve(Util.encode(list.toByteArray()));
                }
                catch (final Exception e) {
                    promise.reject("list", e);
                }
            }
        });
    }

    @ReactMethod
    public void content(final String hash, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                Textile.instance().files.content(hash, new Handlers.DataHandler() {
                    @Override
                    public void onComplete(final byte[] data, final String media) {
                        promise.resolve(Util.encodeData(data, media));
                    }

                    @Override
                    public void onError(final Exception e) {
                        promise.reject("content", e);
                    }
                });
            }
        });
    }

    @ReactMethod
    public void imageContentForMinWidth(final String pth, final Integer minWidth, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                Textile.instance().files.imageContentForMinWidth(pth, minWidth, new Handlers.DataHandler() {
                    @Override
                    public void onComplete(final byte[] data, final String media) {
                        promise.resolve(Util.encodeData(data, media));
                    }

                    @Override
                    public void onError(final Exception e) {
                        promise.reject("imageContentForMinWidth", e);
                    }
                });
            }
        });
    }
}
